crates/runtime/src/instance/allocator/pooling/memory_pool.rs

```diff
@@ -87,78 +170,128 @@ impl MemoryPool {
+            }
+            AutoEnabled::Enable => {
+                if mpk::is_supported() {
+                    mpk::keys()
+                } else {
+                    bail!("mpk is disabled on this system")
+                }
+            }
+            AutoEnabled::Disable => &[],
+        };
+
+        // This is a tricky bit of global state: when creating a memory pool
+        // that uses memory protection keys, we ensure here that any host code
+        // will have access to all keys (i.e., stripes). It's only when we enter
+        // the WebAssembly guest code (see `StoreInner::call_hook`) that we
+        // enforce which keys/stripes can be accessed. Be forewarned about the
+        // assumptions here:
+        // - we expect this "allow all" configuration to reset the default
+        //   process state (only allow key 0) _before_ any memories are accessed
+        // - and we expect no other code (e.g., host-side code) to modify this
+        //   global MPK configuration
+        if !pkeys.is_empty() {
+            mpk::allow(ProtectionMask::all());
+        }
+
         // Interpret the larger of the maximal size of memory or the static
         // memory bound as the size of the virtual address space reservation for
         // memory itself. Typically `static_memory_bound` is 4G which helps
         // elide most bounds checks in wasm. If `memory_pages` is larger,
-        // though, then this is a non-moving pooling allocator so create larger
-        // reservations for account for that.
-        let memory_size = config.limits.memory_pages.max(tunables.static_memory_bound)
+        // though, then this is a non-moving pooling allocator so we create
+        // larger reservations to account for that.
+        let max_memory_bytes = config.limits.memory_pages.max(tunables.static_memory_bound)
             * u64::from(WASM_PAGE_SIZE);
 
-        let memory_and_guard_size =
-            usize::try_from(memory_size + tunables.static_memory_offset_guard_size)
-                .map_err(|_| anyhow!("memory reservation size exceeds addressable memory"))?;
-
-        assert!(
-            memory_and_guard_size % crate::page_size() == 0,
-            "memory size {} is not a multiple of system page size",
-            memory_and_guard_size
-        );
-
-        let max_total_memories = config.limits.total_memories as usize;
-        let initial_memory_offset = if tunables.guard_before_linear_memory {
-            usize::try_from(tunables.static_memory_offset_guard_size).unwrap()
-        } else {
-            0
+        // Create a slab layout and allocate it as completely inaccessible
+        // region to start--`PROT_NONE`.
+        let constraints = SlabConstraints {
+            max_memory_bytes: max_memory_bytes as usize,
+            num_memory_slots: config.limits.total_memories as usize,
+            num_pkeys_available: pkeys.len(),
+            guard_bytes: tunables.static_memory_offset_guard_size as usize,
```

Comment:
Generally I'm pretty wary of `as` casts related to memory code. It's not really an issue for 64-bit platforms but it's one where it can be an issue for 32-bit platforms possibly. Could the fields in `SlabConstraints` have the same types as the inputs here to avoid casts though?

Reply:
Well, it is very convenient to (1) have all these values use the same type so we can do the arithmetic directly without distracting conversions and (2) so we can interact with all the `std` stuff that just uses `usize`. Since they at some point need to be converted to the same type to do the arithmetic, would you prefer if I turned them all into `u64` here and then made conversions to `usize` in all the places we interact with slices/ranges/etc.?

Reply:
Alternately, I could just panic here with `.try_into().unwrap()`? (Or return an error... maybe 32-bit systems don't want to deal with memory sizes they can't address?)

Reply:
`usize::try_from(...).unwrap()` would be my preferred route, if conversions are necessary and we can't just make them go away.